docs/spec/escrow.rst

```diff
@@ -0,0 +1,312 @@
+
+Data Model
+~~~~~~~~~~
+
+Constants
+---------
+
+.. _span:
+
+Span
+....
+
+Used to round heights.
+
+.. _maxPeriod:
+
+MaxPeriod
+.........
+
+The maximum period for lockup.
+
+Scalars
+-------
+
+.. _escrow-scalar-epoch:
+
+Epoch
+.....
+
+The current global epoch for ``PointHistory``.
+
+Maps
+----
+
+.. _escrow-map-locked:
+
+Locked
+......
+
+Stores the ``amount`` and ``end`` block for the account's lock.
+
+.. _escrow-map-point-history:
+
+PointHistory
+............
+
+Stores the global ``bias``, ``slope`` and ``ts`` at a particular point in history,
+
+.. _escrow-map-user-point-history:
+
+UserPointHistory
+................
+
+Stores the ``bias``, ``slope`` and ``ts`` for an account at a particular point in history,
+
+.. _escrow-map-user-point-epoch:
+
+UserPointEpoch
+..............
+
+Stores the current epoch for an account.
+
+.. _escrow-map-slope-changes:
+
+SlopeChanges
+............
+
+Stores scheduled changes of slopes for ending locks.
+
+Structs
+-------
+
+LockedBalance
+.............
+
+The ``amount`` and ``end`` height for a locked balance.
+
+.. tabularcolumns:: |l|L|
+
+==============  ============  ========================================================	
+Parameter       Type          Description                                            
+==============  ============  ========================================================
+``amount``      Balance       The amount deposited to receive vote-escrowed tokens.
+``end``         BlockNumber   The end height after which the balance can be unlocked.
+==============  ============  ========================================================
+
+Point
+.....
+
+The ``bias``, ``slope`` and ``ts`` for our linear function.
+
+.. tabularcolumns:: |l|L|
+
+==============  ============  ========================================================	
+Parameter       Type          Description                                            
+==============  ============  ========================================================
+``bias``        Balance       The bias for the linear function.
+``slope``       Balance       The slope for the linear function.
+``ts``          BlockNumber   The current block height when this point was stored.
+==============  ============  ========================================================
+
+Functions
+~~~~~~~~~
+
+.. _escrow-function-create-lock:
+
+create_lock
+-----------
+
+Create a lock on the account's balance to expire in the future.
+
+Specification
+.............
+
+*Function Signature*
+
+``create_lock(who, amount, unlock_height)``
+
+*Parameters*
+
+* ``who``: The user's address.
+* ``amount``: The amount to be locked.
+* ``unlock_height``: The height to lock until.
+
+*Events*
+
+* :ref:`escrow-event-deposit`
+
+*Preconditions*
+
+* The function call MUST be signed by ``who``.
+* The ``amount`` MUST be non-zero.
+* The account ``who`` MUST NOT already have locked balance.
+* The ``unlock_height`` MUST be greater than ``now``.
```

Comment:
Question: do we allow for arbitrary lock times? IIRC Curve has a fixed period like 2,3, and 4 years. Ok with both, more of me being curious.

Reply:
That's just the Curve UI. The minimum lock is 1 week and they measure is 1 week steps (round down to weeks): https://github.com/curvefi/curve-dao-contracts/blob/master/contracts/VotingEscrow.vy#L419

I think doing something like this makes sense --> makes it less gameable. 
Example: Greg locks for 1 year. Dom sees this and locks for 1 year + 1 block - and beats Greg --> although arguable there is no real difference in terms of stake. 

1 week is not a huge difference, but it's at least not on a "per-block" level. 

I suggest we do 
-  the same as Curve and round-down to weeks (min lock = 1 week)
- Set max time to 2x the parachain lease period (48 weeks * 2). 

--> then we only need to change the "lease period" var for the Polkadot deployment

Reply:
**Lock time measurement**

> Example: Greg locks for 1 year. Dom sees this and locks for 1 year + 1 block - and beats Greg --> although arguable there is no real difference in terms of stake.

The same holds for 1 year + 1 week -> still beat Greg but the lock is just marginally longer. I don't think that is enough of an argument to justify the additional edge case handling over just using blocks directly.

**Lock time = 2x parachain lease period**

We can do that, but I don't think we have that constant on the parachain thus far. @gregdhill you know that better than I do though :) I think it will need to be added for the vesting as well so we might as well have a constant somewhere. My only comment is that the maximum will be different between Kintsugi and Interlay. So any code consuming this, can't take shortcuts.

Reply:
The maximum lock time is already a constant and can be adjusted for Kintsugi and Interlay.